structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py

```diff
@@ -0,0 +1,224 @@
+                    yield
+                finally:
+                    # set the requires_grad back to false
+                    for param in module.parameters():
+                        param.requires_grad = False
+            else:  # if there is no loss_fn, we have nothing to do.
+                warnings.warn(
+                    (
+                        "There is inference_nn in gradient based inference sampler."
+                        " Are you using the right sampler?"
+                    )
+                )
+                try:
+                    yield
+                finally:
+                    pass
+
+    def forward(
+        self,
+        x: Any,
+        labels: Optional[
+            torch.Tensor
+        ] = None,  #: If given will have shape (batch, ...)
+        buffer: Dict = None,
+        **kwargs: Any,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+
+        if labels is None:
+            y_inf: torch.Tensor = self.inference_nn(x, buffer)  # (batch_size, ...)
+            return y_inf, None
+
+        if buffer is None:
+            buffer = {}
+        # switch on gradients on parameters using context manager
+        with self.param_grad():
+
+            loss_fn = self.get_loss_fn(
+                x, labels
+            )  #: Loss function will expect labels in form (batch, num_samples or 1, ...)
+
+            loss_values: List[float] = []
+            loss_values_tensors: List[torch.Tensor] = []
+            step_number = 0
+            loss_value: Union[torch.Tensor, float] = float("inf")
+
+            # we need to enable grad because if the top-level model
+            # was being called in a validation loop, the training
+            # flag will be False for all modules. This will not allow
+            # gradient based inference to progress.
+            with torch.enable_grad():
+                while not self.stopping_criteria(step_number, float(loss_value)):
+                    y_inf: torch.Tensor = self.inference_nn(x, buffer)  # (batch_size, ...)
+
+                    labels = labels.unsqueeze(1)
+                    y_inf = y_inf.unsqueeze(1)
+
+                    y_cost_aug = self.cost_augmented_layer(torch.cat((y_inf, labels), dim=2))  # (batch_size, ...)
+                    loss_value = self.update(
+                        y_inf, y_cost_aug, buffer, loss_fn
+                    )
+
+                    loss_values.append(float(loss_value))
+
+                    step_number += 1
+
+                # buffer['y_inf'] = y_inf
+
+        return y_cost_aug, y_inf
+
+    def update(
+        self,
+        y_hat: torch.Tensor,
+        y_cost_aug: torch.Tensor,
+        buffer: Dict,
+        loss_fn: Callable[[torch.Tensor, torch.Tensor, Dict], torch.Tensor],
+    ) -> torch.Tensor:
+        """
+        Returns:
+            loss: loss value at the previous point (unreduced)
+        """
+        loss = loss_fn(y_hat, y_cost_aug, buffer)
```

Comment:
Before you compute loss, you need to call `optimizer.zero_grad(set_to_none=True)` to zero out the grads. Ref: https://pytorch.org/docs/stable/optim.html#optimizer-step

Reply:
using set_to_none=True is faster than zero_grad(False).

Reply:
In fact, I think you should call zero_grad before any forward steps are done. So perhaps on line 190.